### ISSUE TITLE
Method "as_child" overrides class "Parent" in an incompatible manner

### ISSUE THREAD

**Issue by cleoold** [microsoft/pyright#1824]
**Describe the bug**
Not sure whether it is a bug. But I have two classes: `Parent` and `Child` placed in separate files. The parent has a method that returns the subclass and the subclass now overrides that method. It is an error.

- If I remove child.pyi then no such error reported (removing both has import cycle error)
- If I remove all `TypeVar` and `Generic`s then no error reported

**To Reproduce**
Create folder and paste code below.

**Expected behavior**
?

**Screenshots or Code**
parent.py
```py
from __future__ import annotations
from typing import Generic, TYPE_CHECKING, TypeVar
if TYPE_CHECKING:
    from .child import Child

T = TypeVar('T')

class Parent(Generic[T]):
    def as_child(self) -> Child[T]: ...
```
parent.pyi
```py
from .child import Child as Child
from typing import Generic, TypeVar

T = TypeVar('T')

class Parent(Generic[T]):
    def as_child(self) -> Child[T]: ...

```
child.py
```py
from __future__ import annotations
from typing import TypeVar
from .parent import Parent

T = TypeVar('T')

class Child(Parent[T]):
    def as_child(self) -> Child[T]:  # error
        return self
```
child.pyi
```py
from .parent import Parent
from typing import TypeVar

T = TypeVar('T')

class Child(Parent[T]):
    def as_child(self) -> Child[T]: ...
```

The error says:
```
error: Method "as_child" overrides class "Parent" in an incompatible manner
    Return type mismatch: base method returns type "Child[T@Child]", override returns type "Child[T@Child]"
      "Child[T@Child]" is incompatible with "Child[T@Child]" (reportIncompatibleMethodOverride)
```

**VS Code extension or command-line**
on npm version: 1.1.136

**Additional context**

**Comment by erictraut** [comment]
I see a couple of problems here.

First, why does a parent class have knowledge of one of its children? That's generally a big no-no in object oriented design. I recommend redesigning your class to avoid this.

Second, why are you using type stubs here? The error is due to the fact that you are combining source files with type stubs. Type stubs are meant to be stand-ins for the code in cases where the modules are written in a language other than Python or when the code is imported from another package that is not annotated.

**Comment by cleoold** [comment]
I think the language should still allow the use cases even though something is of bad design.

2. I removed type annotations from the py files, there are still errors saying:
```py
Method "as_child" overrides class "Parent" in an incompatible manner
  Return type mismatch: base method returns type "Child[Unknown]", override returns type "Child"
    "Child" is incompatible with "Child[Unknown]"
```

I sometimes think we have to use both py and stub files. For example, an annotated py file that defines a `__getattr__` and a stub file that has a lot of run-time generated methods. The code example in question is just a minimal reproduction of the problem with irrelevant info omitted.

**Comment by erictraut** [comment]
> I think the language should still allow the use cases even though something is of bad design.

The Python language allows for many bad designs, but that doesn't mean a static type checker needs to accommodate all such designs. Static type checking imposes some constraints on a program. If you want static type checking to work, you sometimes need to replace bad designs or language usage with something better.

Here's the problem with mixing type stubs and source files. The symbol `Child` referenced within the `child.py` file resolves to the `Child` class defined within that same file. The symbol `Child` referenced within the `Parent` class resolves to the `Child` class declared within the `child.pyi` stub file. They are treated as separate types by the type checker. They are not identical things.

You shouldn't use both py and stub files for the same files in the same project.

**Comment by cleoold** [comment]
> Here's the problem with mixing type stubs and source files. The symbol Child referenced within the child.py file resolves to the Child class defined within that same file. The symbol Child referenced within the Parent class resolves to the Child class declared within the child.pyi stub file. They are treated as separate types by the type checker. They are not identical things.

Why is there no error if I remove generics from the sources?
